safe-transaction-client/src/lib.rs

```diff
@@ -0,0 +1,343 @@
+        use ethers::abi::Tokenizable;
+        use ethers::utils::keccak256;
+        use tiny_keccak::{Hasher, Keccak};
+        macro_rules! tokenize {
+            ($($tok:expr),* $(,)*) => {
+                &[$($tok.into_token()),*][..]
+            }
+        }
+
+        let safe_tx_hash = TxHash::from({
+            let mut hasher = Keccak::v256();
+            let mut output = [0u8; 32];
+            hasher.update(&[0x19, 0x01]);
+            // domain separator
+            hasher.update(&keccak256(&ethers::abi::encode(tokenize![
+                // DOMAIN_SEPARATOR_TYPEHASH
+                U256::from_str(
+                    "0x47e79534a245952e8b16893a336b85a3d9ea9fa8c573f3d803afb92a79469218"
+                )
+                .unwrap(),
+                U256::from(signer.chain_id()),
+                self.safe_address,
+            ])));
+            // safeTxHash
+            hasher.update(&keccak256(&ethers::abi::encode(tokenize![
+                // SAFE_TX_TYPEHASH
+                U256::from_str(
+                    "0xbb8310d486368db6bd6f849402fdd73ad53d316b5a4b2644ad6efe0f941286d8"
+                )
+                .unwrap(),
+                self.to,
+                self.value,
+                keccak256(&self.data),
+                self.operation as u8,
+                U256::zero(),    // safe_tx_gas
+                U256::zero(),    // base_gas
+                U256::zero(),    // gas_price
+                Address::zero(), // gas_token
+                Address::zero(), // refund_receiver
+                self.nonce,
+            ])));
+            hasher.finalize(&mut output);
+            output
+        });
+
+        let SignedSafeTxHash { signature, .. } = sign_tx_hash(signer, safe_tx_hash).await?;
+
+        Ok(SignedSafeTx {
+            inner: self,
+            sender: signer.address(),
+            safe_tx_hash,
+            signature,
+        })
+    }
+}
+
+pub struct SignedSafeTxHash {
+    safe_tx_hash: TxHash,
+    signature: Signature,
+}
+
+pub async fn sign_tx_hash<S>(
+    signer: &S,
+    safe_tx_hash: TxHash,
+) -> Result<SignedSafeTxHash, Error<S::Error>>
+where
+    S: Signer,
+    Error<S::Error>: From<S::Error>,
+{
+    let mut signature = signer.sign_message(safe_tx_hash).await?;
+
+    // signature.recover expects signature.v ∈ {27, 28}
+    if signature.v < 2 {
+        signature.v += 27;
+    }
+    assert!(
+        [27, 28].contains(&signature.v),
+        "BUG: `Signer::sign_message` produced an invalid `v` field."
+    );
+    assert_eq!(
+        Some(signer.address()),
+        signature.recover(&safe_tx_hash[..]).ok(),
+        "BUG: `Signature::recover` failed to recover our signing \
+            address... did `Signer::sign_message` fail to add the prefix?"
+    );
+    signature.v += 4;
+
+    Ok(SignedSafeTxHash {
+        safe_tx_hash,
+        signature,
+    })
+}
+
+#[test]
+#[cfg(feature = "run-test-against-api")]
+fn check() {
+    futures_executor::block_on(async {
```

Comment:
This makes me lol

Reply:
Sanity check. We can't write `#[test] async fn check() { .. }`. It is certainly not worth depending on `tokio` just for `#[tokio::test]`. :)

Reply:
Yeah kind of weird there isn't a `futures::test` decorator ...